core/src/com/tgf/twf/core/world/storage/HashMapInventory.java

```diff
@@ -28,56 +27,39 @@ public int getTotalStoredQuantity() {
     }
 
     @Override
-    public void store(final ResourceType resourceType, final int quantity) {
+    public boolean store(final ResourceType resourceType, final int quantity) {
         stock.put(resourceType, stock.getOrDefault(resourceType, 0) + quantity);
         totalStoredQuantity += quantity;
+        return true;
     }
 
     @Override
-    public void store(final Inventory inventory) {
+    public boolean store(final Inventory inventory) {
         for (final ResourceType resourceType : inventory.getStoredResourceTypes()) {
             store(resourceType, inventory.getStoredQuantity(resourceType));
         }
+        return true;
     }
 
-    @Override
-    public void transfer(final Storage destinationStorage) {
-        final Set<ResourceType> emptyResourceTypes = new HashSet<>();
-        stock.forEach((key, value) -> {
-            final int actuallyStored = destinationStorage.store(key, value);
-            final boolean isEmpty = retrieve(key, actuallyStored, false);
-            if (isEmpty) {
-                emptyResourceTypes.add(key);
-            }
-        });
-        emptyResourceTypes.forEach(stock::remove);
-    }
 
     @Override
-    public void retrieve(final ResourceType resourceType, final int quantity) {
-        retrieve(resourceType, quantity, true);
-    }
-
-    @Override
-    public void clear() {
-        stock.clear();
-    }
-
-    /**
-     * @return true if there are no resources of the specified type left
-     */
-    private boolean retrieve(final ResourceType resourceType, final int quantity, final boolean removeEntry) {
+    public boolean retrieve(final ResourceType resourceType, final int quantity) {
         final int currentStock = stock.getOrDefault(resourceType, 0);
         final int newStock = currentStock - quantity;
         if (newStock < 0) {
             throw new IllegalStateException("Cannot retrieve " + quantity + " of " + resourceType + "; only " + currentStock + " available");
         }
-        if (newStock == 0 && removeEntry) {
```

Comment:
Why do we not care about this flag anymore?

Reply:
This was necessary because how transfer was implemented. Initially we iterated for each resource type, and tried to transfer it to the destination storage, and if after that the remaining stock was empty we removed it. Removing in place caused a concurrent modification exception.

Now because operations are either completely succeeding or failing, there's no such complexity anymore. If an inventory transfer succeeds, we can just completely clear the inventory after the transfer. See snippet below. Note that in the PR there's a bug that I only picked later, there's no clear call.

    /**
     * Transfer the whole inventory to the destination storage.
     *
     * @return true if the inventory was transferred, depending on destination storage capacity.
     */
    public boolean transfer(final Storage destinationStorage) {
        if (inventory.getTotalStoredQuantity() == 0) {
            return true;
        }
        final boolean canStore = inventory.getStoredResourceTypes().stream()
                .allMatch(resourceType -> destinationStorage.getRemainingCapacity(resourceType) >= inventory.getStoredQuantity(resourceType));
        if (canStore) {
            destinationStorage.store(inventory);
            inventory.clear();
        }
        return canStore;
    }